### PR TITLE
#175494065 Ft reset password

### PR DESCRIPTION
**Pull request by BenieSereine**
#### What does this PR do?
This PR reset the password of the user
#### Description of Task to be completed?
- User receive an email that helps him/her to reset her password
- User is able to reset the password when forgotten
#### How should this be manually tested?
- clone the repo using `git clone https://github.com/Soma-Technologies-Inc/visitAfrica-backend.git`
- run `git checkout ft-resetPassword` to navigate to my branch of reset Password
- run `git pull ` to make sure you have updated code
- run `npm install`
- run `npm run migrate` to create migrations for the database
- run `npm run startDev` to run the server
- First signup to receive the token to be used.
- Open postman and using post `http://localhost:3000/api/auth/forgetpassword` to send an email to the user in the body using the token gotten typing `{
    "email": "BMugwaneza720@daviscollege.com"
}` and the token received you use it to update the password and patch `http://localhost:3000/api/auth/resetpassword` with the information `{
    "password":"Benie@123",
    "confirmPassword":"Benie@123"
}` to change the password.
- There must be Authorization in the header to change the password.
#### Any background context you want to provide?
N/A
#### forgetting the password
Screenshot from 2020-12-15 15-27-29
### email sent
Screenshot from 2020-12-16 13-57-47
#### resetting
Screenshot from 2020-12-16 13-58-28
#### when password updated
Screenshot from 2020-12-16 13-54-14
#### when you want to reset the password and put unmatched password
Screenshot from 2020-12-15 15-29-26
#### Then login using the changed password
Screenshot from 2020-12-18 13-50-49

### CODE REVIEW

**Comment by tigthor** [comment]
Please work on  your pr description and how this can be tested .....
as this is incomplete

File: src/controllers/auth.controller.js [hunk 1]
```diff
@@ -93,6 +93,54 @@ class AuthController {
...
+ * @param {object} res
+ * @returns {object} resetting the password
+*/
+	static async resetPassword(req, res) {
+		const decodedToken = await TokenService.verifyToken(req.query.token);
+		await UserService.updateUserByAttribute({ email: decodedToken.email }, {
+			password: BcryptService.hashPassword(req.body.password),
+			confirmPassword: BcryptService.hashPassword(req.body.confirmPassword)
+		});
+		if (req.body.password !== req.body.confirmPassword) {
```
Comment (tigthor) [inline src/controllers/auth.controller.js:111]: You should put this if condition in the middleware.

File: src/controllers/auth.controller.js [hunk 2]
```diff
@@ -93,6 +93,54 @@ class AuthController {
...
+		ResponseService.setSuccess(200, 'Password changed successfully');
+		return ResponseService.send(res);
+	}
+
+	/**
+	 * @param {object} req
+	 * @param {object} res
+	 * @returns {object} helping the user to reset the password
+	*/
+	static async resetpass(req, res) {
```
Comment (tigthor) [inline src/controllers/auth.controller.js]: Your naming convetion take an example you named the async function on line 182 resetPassword and this line of 121 resetpass please use better naming convetion of your functions i the controller

File: src/controllers/auth.controller.js [hunk 3]
```diff
@@ -93,6 +93,54 @@ class AuthController {
...
+*/
+	static async sendResetPasswordLink(req, res) {
+		const result = await UserService.findUserByAttribute({ email: req.body.email });
+		if (result !== null) {
+			const token = TokenService.generateToken({ email: req.body.email });
+			MailService.sendEMail(result.dataValues.fullname, result.dataValues.email, token);
+			ResponseService.setSuccess(200, 'Email sent, please check your email to reset your password', token);
+			return ResponseService.send(res);
+		}
+		ResponseService.setError(404, 'user not found!');
```
Comment (tigthor) [inline src/controllers/auth.controller.js:144]: Please use a middleware to first check if a user exist then send this response... You should not put this in the controller

File: src/services/mail.service.js [hunk 4]
```diff
@@ -47,6 +47,49 @@ class MailService {
...
+			product: {
+				name: 'Visit Africa App',
+				link: `${process.env.BASE_URL}/api/auth/resetpassword?token=${token}`,
+			},
+		});
+		const emailTemplate = mailGenerator.generate(emailToSend);
+		require('fs').writeFileSync('preview.html', emailTemplate, 'utf8');
+		const msg = {
+			to: email,
+			from: 't0780586360@gmail.com',
```
Comment (tigthor) [inline src/services/mail.service.js]: Change this email into yours

**Review by tigthor — CHANGES_REQUESTED** [review]
Please work on my comments

**Review by tigthor — APPROVED** [review]

File: src/routes/auth.route.js [hunk 5]
```diff
@@ -46,5 +48,7 @@ router.get(
 	}),
 	AuthController.loginWithSocialMedia
 );
+router.patch('/reset-password', protectroute, resetPassword, AuthController.resetPassword);
```
Comment (Niyonsengaeric) [inline src/routes/auth.route.js:51]: what are you protecting here?

File: src/tests/auth/resetpassword.test.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+const userVerificationTest = () => {
+	describe('/POST forget password', () => {
+		before(() => {
+			createUser(forgetPassword);
+		});
+		it('Should help a user who forgot the password', (done) => {
+			chai.request(app)
+				.post('/api/auth/forget-password')
+				.set('Authorization', `Bearer ${forgotPasswordToken}`)
```
Comment (Niyonsengaeric) [inline src/tests/auth/resetpassword.test.js]: why are you still sitting authorization?

File: src/validations/user.validation.js [hunk 7]
```diff
@@ -193,3 +193,48 @@ export const validateUserBody = (req, res, next) => {
...
+		password: Joi.string()
+			.regex(/^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\$%\^&\*])(?=.{8,})/)
+			.required()
+			.messages({
+				'string.pattern.base': 'password should contain uppercase,lowercase,specialCharacter,and number',
+				'any.required': 'password is required',
+				'string.empty': 'password is not allowed to be empty'
+
+			}),
+		confirmPassword: Joi.string()
```
Comment (Niyonsengaeric) [inline src/validations/user.validation.js]: can you check on the way of validating if it is equal to password without setting again this policies?

File: src/controllers/auth.controller.js (hunk 1 above)
Comment (BenieSereine) [reply src/controllers/auth.controller.js:111]: done

File: src/controllers/auth.controller.js (hunk 2 above)
Comment (BenieSereine) [reply src/controllers/auth.controller.js]: done

File: src/controllers/auth.controller.js (hunk 3 above)
Comment (BenieSereine) [reply src/controllers/auth.controller.js:144]: done

File: src/services/mail.service.js (hunk 4 above)
Comment (BenieSereine) [reply src/services/mail.service.js]: done

File: src/routes/auth.route.js (hunk 5 above)
Comment (BenieSereine) [reply src/routes/auth.route.js:51]: done

File: src/tests/auth/resetpassword.test.js (hunk 6 above)
Comment (BenieSereine) [reply src/tests/auth/resetpassword.test.js]: done

File: src/validations/user.validation.js (hunk 7 above)
Comment (BenieSereine) [reply src/validations/user.validation.js]: done

**Pull request merged**